File: modular_exponential/modular_exponential.cs
using System;

public class modular_exponential
{
	public static int modularExponential(int b, int power, int modulus)
	{
		if (b > modulus)
			b %= modulus;
		
		int result = 1;
		while (power > 0)
		{
			if (power % 2 == 1)
				result = (result * b) % modulus;
			power = power >> 1;
			b = (b * b) % modulus;
		}
		return result;
	}

	static void Main(string[] args)
	{
		Console.WriteLine(modularExponential(24, 2015, 1000));
	}
}

File: modular_exponential/modular_exponential.cpp
#include <iostream>
using namespace std;

int modular_exponential(long base, long power, int modulus) {

	if (base >= modulus)
		base %= modulus;

	int result = 1;

	while (power > 0) {
		if (power & 1)
			result = (result * base) % modulus;
		power = power >> 1;
		base = (base * base) % modulus;
	}
	return result;
}

int main() {

	cout << modular_exponential(24, 2015, 1000) << endl;
	return 0;
}
